src/flint/types/fq_default.pyx

```diff
@@ -0,0 +1,933 @@
+
+        # If other is not an fq_default element, we attempt to convert to fq_default
+        if not typecheck(other, fq_default):
+            # For nmod and fmpz_mod if the modulus does not match the characteristic
+            # then we return false.
+            if typecheck(other, nmod) and self.ctx.characteristic() != (<nmod>other).modulus():
+                    res = False
+
+            elif typecheck(other, fmpz_mod) and self.ctx.characteristic() != (<fmpz_mod>other).ctx.modulus():
+                    res = False
+
+            else:
+                # Convert from int, fmpz, fmpz_mod and nmod to fq_default
+                cmp = self.ctx.new_ctype_fq_default()
+                check = self.ctx.set_any_scalar_as_fq_default((<fq_default>cmp).val, other)
+
+                # Conversion failed, element was not a compatible scalar
+                if check is NotImplemented:
+                    res = False
+                else:
+                    # We now have an fq_default element with the same context, so compare value only
+                    res = fq_default_equal(self.val, (<fq_default>cmp).val, self.ctx.val)
+
+            # Flip the result of res if we're doing not equals
+            if op == 2:
+                return res
+            return not res
+
+        # Otherwise we're in the case where other is also an fq_default but may have a
+        # different context.
+
+        # If the contexts match exactly, only check values
+        if self.ctx == (<fq_default>other).ctx:
+            res = fq_default_equal(self.val, (<fq_default>other).val, self.ctx.val)
+
+        # Otherwise if both contexts have the same characteristic check
+        # if both fq_default lift to the same integer
+        elif self.ctx.characteristic() == (<fq_default>other).ctx.characteristic():
+            try:
+                res = int(self) == int(other)
+            # One or both lifts failed, so values are not equal
+            except ValueError:
+                res = False
+
+        # Otherwise, values are considered not equal
+        else:
+            res = False
+
+        # Flip the result of res if we're doing not equals
+        if op == 2:
+            return res
+        return not res
+
+    # =================================================
+    # Generic arithmetic required by flint_scalar
+    # =================================================
+
+    def _any_as_self(self, other):
+        return self.ctx.any_as_fq_default(other)
+
+    def _neg_(self):
+        cdef fq_default res
+        res = self.ctx.new_ctype_fq_default()
+        fq_default_neg(res.val, self.val, self.ctx.val)
+        return res
+
+    def _add_(self, other):
+        """
+        Assumes that __add__() has ensured other is of type self
+        """
+        cdef fq_default res
+        res = self.ctx.new_ctype_fq_default()
+        fq_default_add(res.val, self.val, (<fq_default>other).val, self.ctx.val)
+        return res
```

Comment:
I would add types for the parameters here and make this cdef I think:
```python
cdef fq_default _add_(fq_default self, fq_default other):
    cdef fq_default res = self.ctx.new_ctype_fq_default()
    fq_default_add(res.val, self.val, other.val, self.ctx.val)
    return res
```

Reply:
I'm not sure if making this `cdef` makes much difference though.

Reply:
I tried doing this and it broke all the arithmetic... Is there a non-trivial change needed to be made to the base class too?

Reply:
I'm not sure what is the best way to make this work. If we want to use this widely though (for other types) then it would be good to benchmark, look at the generated C code, try alternatives etc.

I am concerned about the overhead of these methods. See e.g.: https://github.com/sympy/sympy/discussions/26932. I got 10% speedup on a macro benchmark that used fmpz just by inlining the type checking in `__add__`:
```diff
diff --git a/src/flint/types/fmpz.pyx b/src/flint/types/fmpz.pyx
index 0b16375..0a025b8 100644
--- a/src/flint/types/fmpz.pyx
+++ b/src/flint/types/fmpz.pyx
@@ -1,3 +1,5 @@
+from cpython.object cimport PyTypeObject, PyObject_TypeCheck
+
 from flint.flint_base.flint_base cimport flint_scalar
 from flint.utils.typecheck cimport typecheck
 from flint.utils.conversion cimport chars_from_str
@@ -185,6 +187,10 @@ cdef class fmpz(flint_scalar):
         return -self
 
     def __add__(s, t):
+        if PyObject_TypeCheck(t, <PyTypeObject*>fmpz):
+            u = fmpz.__new__(fmpz)
+            fmpz_add((<fmpz>u).val, (<fmpz>s).val, (<fmpz>t).val)
+            return u
         cdef fmpz_struct tval[1]
         cdef int ttype = FMPZ_UNKNOWN
         u = NotImplemented
```
It still came out notably slower than when using gmpy2 though so there is definitely some overhead in the Cython wrapper that could be reduced.

Reply:
Maybe we should just prioritise speed for all these scalar types and accept we have to hand write all the __add__ methods etc. seems silly to lose a bunch of performance just for a few smaller functions.

Reply:
We need to measure the effects. I think it is fine to have a framework in the superclass but then inline common cases. Note that e.g. for SymPy the case that really matters is the common case like `fmpz*fmpz` and to a much lesser extent `fmpz*int`. We could have an `fmpz.__add__` method that optimises that common case as hard as possible but then just falls back on something slower that handles coercions etc:
```python
class fmpz(scalar):
    def __add__(self, other):
        if PyTypeCheck(...):
            # Fast optimised case
        else:
            # Fall back on slower generic routine
            return super().__add__(self, other)
```
It makes sense to start out having a nice generic routine in the superclass that handles everything correctly and then after doing timing measurements micro-optimise the cases that matter.

Reply:
OK. We can work this way then. I think int * type and type * type are the ones worth doing first as I think it'll be comment for people to do things like 5 * type or type + 1 in their code and it would be nice for this to be fast without having to do one = type(one); type + one

Reply:
Exactly and almost always the ints would be small so the fast path could use `fmpz_mul_si` and friends rather than general `int -> fmpz` coercion.